File: transports/java/src/test/java/community/catalyst/csilgen/transport/ConformanceTest.java
// Verify the Java transport library against the checked-in conformance vectors.
//
// This both guards the encoders/decoders against drift and demonstrates the contract every
// reference library follows: reconstruct each vector's envelope from its language-neutral
// `input`, assert encode → `hex`, and assert decode(`hex`) → the same envelope. Ported from
// transports/go/conformance_test.go and transports/rust/tests/conformance.rs.
package community.catalyst.csilgen.transport;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

class ConformanceTest {

    static List<Vectors.Case> rpcVectors() {
        return Vectors.load("rpc.json");
    }

    static List<Vectors.Case> eventVectors() {
        return Vectors.load("events.json");
    }

    static List<Vectors.Case> datagramVectors() {
        return Vectors.load("datagrams.json");
    }

    @ParameterizedTest(name = "rpc/{0}")
    @MethodSource("rpcVectors")
    void rpc(Vectors.Case vec) {
        Map<String, Object> in = vec.input();
        byte[] actual;
        switch (Vectors.reqStr(in, "kind")) {
            case "request" -> {
                Rpc.Request r = new Rpc.Request(
                        Vectors.reqStr(in, "service"),
                        Vectors.reqStr(in, "op"),
                        Vectors.optLong(in, "id"),
                        Vectors.hex(Vectors.reqStr(in, "payload_hex")),
                        Vectors.optStr(in, "auth"));
                actual = r.encode();
                assertEquals(r, Rpc.Request.decode(actual), "decode mismatch " + vec.name());
            }
            case "response" -> {
                Rpc.Response r = new Rpc.Response(
                        Vectors.optLong(in, "id"),
                        Status.fromCode(Vectors.reqLong(in, "status")),
                        Vectors.optStr(in, "variant"),
                        Vectors.optStr(in, "error"),
                        Vectors.hex(Vectors.reqStr(in, "payload_hex")));
                actual = r.encode();
                assertEquals(r, Rpc.Response.decode(actual), "decode mismatch " + vec.name());
            }
            case "push" -> {
                Rpc.Push p = new Rpc.Push(
                        Vectors.reqStr(in, "service"),
                        Vectors.reqStr(in, "event"),
                        Vectors.hex(Vectors.reqStr(in, "payload_hex")));
                actual = p.encode();
                assertEquals(p, Rpc.Push.decode(actual), "decode mismatch " + vec.name());
            }
            default -> throw new IllegalArgumentException("unknown rpc kind");
        }
        assertEquals(vec.hex(), Vectors.toHex(actual), "encode mismatch " + vec.name());
    }

    @ParameterizedTest(name = "events/{0}")
    @MethodSource("eventVectors")
    void events(Vectors.Case vec) {
        Map<String, Object> in = vec.input();
        String control = Vectors.optStr(in, "control");
        if (control != null) {
            byte[] bytes = encodeControl(control, in);
            assertEquals(vec.hex(), Vectors.toHex(bytes), "encode mismatch " + vec.name());
            return;
        }

        Events.Profile profile = Events.Profile.parse(Vectors.reqStr(in, "profile"));
        byte[] payload = Vectors.hex(Vectors.reqStr(in, "payload_hex"));
        Events.Event ev;
        if (profile == Events.Profile.VERBOSE) {
            ev = new Events.Event(Vectors.optStr(in, "service"), null, Vectors.reqStr(in, "event"),
                    null, Vectors.optLong(in, "id"), payload);
        } else {
            Long id = Vectors.optLong(in, "id");
            ev = new Events.Event(null, Vectors.reqLong(in, "service_ord"), null,
                    Vectors.reqLong(in, "op_ord"), id, payload);
        }
        byte[] actual = ev.encode(profile);
        assertEquals(vec.hex(), Vectors.toHex(actual), "encode mismatch " + vec.name());
        assertEquals(ev, Events.Event.decode(actual, profile), "decode mismatch " + vec.name());
    }

    @SuppressWarnings("unchecked")
    private static byte[] encodeControl(String control, Map<String, Object> in) {
        switch (control) {
            case "hello" -> {
                List<Long> versions = new ArrayList<>();
                for (Object x : (List<Object>) in.get("versions")) {
                    versions.add((long) (double) (Double) x);
                }
                List<String> profiles = new ArrayList<>();
                for (Object x : (List<Object>) in.get("profiles")) {
                    profiles.add((String) x);
                }
                return new Events.Hello(versions, profiles, Vectors.optStr(in, "service"),
                        Vectors.optStr(in, "auth")).encode();
            }
            case "hello_ack" -> {
                return new Events.HelloAck(Vectors.reqLong(in, "v"), Vectors.reqStr(in, "profile"),
                        Vectors.optStr(in, "session")).encode();
            }
            case "ping" -> {
                return new Events.Heartbeat(Vectors.reqLong(in, "nonce"), Vectors.optLong(in, "at"))
                        .encode();
            }
            case "close" -> {
                return new Events.Close(Status.fromCode(Vectors.reqLong(in, "status")),
                        Vectors.optStr(in, "reason")).encode();
            }
            default -> throw new IllegalArgumentException("unknown control " + control);
        }
    }

    @ParameterizedTest(name = "datagrams/{0}")
    @MethodSource("datagramVectors")
    void datagrams(Vectors.Case vec) {
        Map<String, Object> in = vec.input();
        switch (Vectors.reqStr(in, "profile")) {
            case "cbor-array" -> {
                Datagrams.Datagram d = new Datagrams.Datagram(Vectors.reqLong(in, "op_ord"),
                        Vectors.reqLong(in, "seq"), Vectors.hex(Vectors.reqStr(in, "payload_hex")));
                byte[] actual = d.encode();
                assertEquals(vec.hex(), Vectors.toHex(actual), "encode mismatch " + vec.name());
                assertEquals(d, Datagrams.Datagram.decode(actual),
                        "decode mismatch " + vec.name());
            }
            case "compact-header" -> {
                Datagrams.CompactDatagram d = Datagrams.CompactDatagram.of(
                        (int) Vectors.reqLong(in, "op_ord"),
                        (int) Vectors.reqLong(in, "seq"),
                        Vectors.hex(Vectors.reqStr(in, "body_hex")));
                Long epoch = Vectors.optLong(in, "epoch");
                if (epoch != null) {
                    d = d.withEpoch(epoch.intValue());
                }
                byte[] actual = d.encode();
                assertEquals(vec.hex(), Vectors.toHex(actual), "encode mismatch " + vec.name());
                assertEquals(d, Datagrams.CompactDatagram.decode(actual),
                        "decode mismatch " + vec.name());
            }
            default -> throw new IllegalArgumentException("unknown datagram profile");
        }
    }
}

File: transports/kotlin/src/test/kotlin/community/catalyst/csilgen/transport/ConformanceTest.kt
// Verify the Kotlin transport library against the checked-in conformance vectors.
//
// This both guards the encoders/decoders against drift and demonstrates the contract every
// reference library follows: reconstruct each vector's envelope from its language-neutral
// `input`, assert encode → `hex`, and assert decode(`hex`) → the same envelope. Ported from
// transports/go/conformance_test.go and transports/rust/tests/conformance.rs.
//
// `data class`/ByteArray identity equality is the headline Kotlin gotcha: the envelope
// classes override equals/hashCode with contentEquals, and the byte-level encode check uses
// assertContentEquals — assertEquals on raw ByteArray would compare references and silently
// pass-or-fail by identity.
package community.catalyst.csilgen.transport

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.fail

private fun loadVectors(name: String): List<Map<String, Any?>> {
    val path = File("../conformance/$name")
    @Suppress("UNCHECKED_CAST")
    val doc = Json.parse(path.readText()) as Map<String, Any?>
    @Suppress("UNCHECKED_CAST")
    return doc["vectors"] as List<Map<String, Any?>>
}

private fun optStr(m: Map<String, Any?>, k: String): String? = m[k] as String?

private fun optU64(m: Map<String, Any?>, k: String): ULong? =
    (m[k] as Double?)?.toLong()?.toULong()

private fun reqStr(m: Map<String, Any?>, k: String): String =
    m[k] as String? ?: fail("missing required string field '$k'")

private fun reqU64(m: Map<String, Any?>, k: String): ULong =
    (m[k] as Double? ?: fail("missing required integer field '$k'")).toLong().toULong()

private fun reqI64(m: Map<String, Any?>, k: String): Long =
    (m[k] as Double? ?: fail("missing required integer field '$k'")).toLong()

/** Assert the encoded bytes match the vector hex, comparing as hex strings so a mismatch
 *  prints a readable diff rather than two byte arrays. */
private fun assertHex(name: String, expected: String, actual: ByteArray) {
    assertEquals(expected, toHex(actual), "$name encode mismatch")
}

class ConformanceTest {
    @Test
    fun rpcVectors() {
        for (vec in loadVectors("rpc.json")) {
            val name = vec["name"] as String
            val hex = vec["hex"] as String
            val input = @Suppress("UNCHECKED_CAST") (vec["input"] as Map<String, Any?>)
            when (val kind = reqStr(input, "kind")) {
                "request" -> {
                    val r = RpcRequest(
                        service = reqStr(input, "service"),
                        op = reqStr(input, "op"),
                        payload = unhex(reqStr(input, "payload_hex")),
                        id = optU64(input, "id"),
                        auth = optStr(input, "auth"),
                    )
                    assertHex(name, hex, r.encode())
                    assertEquals(r, RpcRequest.decode(r.encode()), "$name decode mismatch")
                }
                "response" -> {
                    val r = RpcResponse(
                        status = Status.fromCode(reqI64(input, "status").toInt()),
                        payload = unhex(reqStr(input, "payload_hex")),
                        id = optU64(input, "id"),
                        variant = optStr(input, "variant"),
                        error = optStr(input, "error"),
                    )
                    assertHex(name, hex, r.encode())
                    assertEquals(r, RpcResponse.decode(r.encode()), "$name decode mismatch")
                }
                "push" -> {
                    val p = RpcPush(
                        service = reqStr(input, "service"),
                        event = reqStr(input, "event"),
                        payload = unhex(reqStr(input, "payload_hex")),
                    )
                    assertHex(name, hex, p.encode())
                    assertEquals(p, RpcPush.decode(p.encode()), "$name decode mismatch")
                }
                else -> fail("unknown rpc kind '$kind'")
            }
        }
    }

    @Test
    fun eventsVectors() {
        for (vec in loadVectors("events.json")) {
            val name = vec["name"] as String
            val hex = vec["hex"] as String
            val input = @Suppress("UNCHECKED_CAST") (vec["input"] as Map<String, Any?>)

            val control = optStr(input, "control")
            if (control != null) {
                val bytes = when (control) {
                    "hello" -> Hello(
                        versions = (input["versions"] as List<*>).map { (it as Double).toLong().toULong() },
                        profiles = (input["profiles"] as List<*>).map { it as String },
                        service = optStr(input, "service"),
                        auth = optStr(input, "auth"),
                    ).encode()
                    "hello_ack" -> HelloAck(
                        v = reqU64(input, "v"),
                        profile = reqStr(input, "profile"),
                        session = optStr(input, "session"),
                    ).encode()
                    "ping" -> Heartbeat(
                        nonce = reqU64(input, "nonce"),
                        at = optU64(input, "at"),
                    ).encode()
                    "close" -> Close(
                        status = Status.fromCode(reqI64(input, "status").toInt()),
                        reason = optStr(input, "reason"),
                    ).encode()
                    else -> fail("unknown control '$control'")
                }
                assertHex(name, hex, bytes)
                continue
            }

            val profile = Profile.parse(reqStr(input, "profile")) ?: fail("unknown profile")
            val payload = unhex(reqStr(input, "payload_hex"))
            val ev = when (profile) {
                Profile.VERBOSE ->
                    Event.verbose(optStr(input, "service"), reqStr(input, "event"), payload)
                        .let { if (optU64(input, "id") != null) it.withId(optU64(input, "id")!!) else it }
                Profile.COMPACT ->
                    Event.compact(reqU64(input, "service_ord"), reqU64(input, "op_ord"), payload)
                        .let { if (optU64(input, "id") != null) it.withId(optU64(input, "id")!!) else it }
            }
            assertHex(name, hex, ev.encode(profile))
            assertEquals(ev, Event.decode(ev.encode(profile), profile), "$name decode mismatch")
        }
    }

    @Test
    fun datagramVectors() {
        for (vec in loadVectors("datagrams.json")) {
            val name = vec["name"] as String
            val hex = vec["hex"] as String
            val input = @Suppress("UNCHECKED_CAST") (vec["input"] as Map<String, Any?>)
            when (val profile = reqStr(input, "profile")) {
                "cbor-array" -> {
                    val d = Datagram(
                        opOrd = reqU64(input, "op_ord"),
                        seq = reqU64(input, "seq"),
                        payload = unhex(reqStr(input, "payload_hex")),
                    )
                    assertHex(name, hex, d.encode())
                    assertEquals(d, Datagram.decode(d.encode()), "$name decode mismatch")
                }
                "compact-header" -> {
                    var d = CompactDatagram(
                        opOrd = reqU64(input, "op_ord").toUByte(),
                        seq = reqU64(input, "seq").toUShort(),
                        body = unhex(reqStr(input, "body_hex")),
                    )
                    optU64(input, "epoch")?.let { d = d.withEpoch(it.toUByte()) }
                    assertHex(name, hex, d.encode())
                    assertEquals(d, CompactDatagram.decode(d.encode()), "$name decode mismatch")
                }
                else -> fail("unknown datagram profile '$profile'")
            }
        }
    }

    @Test
    fun everyVectorFileIsNonEmpty() {
        assertTrue(loadVectors("rpc.json").isNotEmpty())
        assertTrue(loadVectors("events.json").isNotEmpty())
        assertTrue(loadVectors("datagrams.json").isNotEmpty())
    }
}
